pallets/xyk/src/lib.rs

```diff
@@ -2067,6 +2217,141 @@ impl<T: Config> PreValidateSwaps for Pallet<T> {
 			sold_asset_amount,
 		))
 	}
+
+	/// We only validate the first atomic swap's ability to accept fees
+	fn pre_validate_multiswap_buy_asset(
+		sender: &Self::AccountId,
+		swap_token_list: Vec<Self::CurrencyId>,
+		final_bought_asset_amount: Self::Balance,
+		max_amount_in: Self::Balance,
+	) -> Result<
+		(
+			Self::Balance,
+			Self::Balance,
+			Self::Balance,
+			Self::Balance,
+			Self::Balance,
+			Self::CurrencyId,
+			Self::CurrencyId,
+		),
+		DispatchError,
+	> {
+		// Ensure not buying zero amount
+		ensure!(!final_bought_asset_amount.is_zero(), Error::<T>::ZeroAmount,);
+		ensure!(!max_amount_in.is_zero(), Error::<T>::ZeroAmount,);
+
+		ensure!(swap_token_list.len() > 2_usize, Error::<T>::MultiswapShouldBeAtleastTwoHops);
+		// Unwraps are fine due to above ensure
+		let sold_asset_id = *swap_token_list.get(0).unwrap();
+		let bought_asset_id = *swap_token_list.get(1).unwrap();
+
+		ensure!(
+			!T::DisabledTokens::contains(&sold_asset_id) &&
+				!T::DisabledTokens::contains(&bought_asset_id),
+			Error::<T>::FunctionNotAvailableForThisToken
+		);
+
+		// Cannot use multiswap twice on the same pool
+		let mut pair_sold_assets = swap_token_list.clone();
+		let _ = pair_sold_assets.pop();
+
+		let mut pair_bought_assets = swap_token_list.clone();
+		let _ = pair_bought_assets.remove(0);
+
+		let atomic_pairs: Vec<(TokenId, TokenId)> =
+			pair_sold_assets.into_iter().zip(pair_bought_assets.into_iter()).collect();
+
+		let mut filtered_atomic_pairs: Vec<(TokenId, TokenId)> = atomic_pairs
+			.into_iter()
+			.filter_map(|(x, y)| {
+				if x == y {
+					None
+				} else if x > y {
+					Some((x, y))
+				} else
+				// x < y
+				{
+					Some((y, x))
+				}
+			})
+			.collect();
+
```

Comment:
if you collect directly to hashset, you can avoide few unecescary allocations, and separate dedup call